tools/spell-check.sh

```diff
@@ -7,55 +7,93 @@ cd "$(dirname "$0")"/..
-            words+=$'\n'
-            words+="${word}"
-        fi
-    done
+    dependencies=$(sed <<<"${dependencies}" 's/[0-9_-]/\n/g' | LC_ALL=C sort -f -u)
+fi
+config_old=$(<.cspell.json)
+config_new=$(grep <<<"${config_old}" -v ' *//' | jq 'del(.dictionaries[])' | jq 'del(.dictionaryDefinitions[])')
+echo "${config_new}" >.cspell.json
+if [[ -n "${has_rust}" ]]; then
+    dependencies_words=$(npx <<<"${dependencies}" cspell stdin --no-progress --no-summary --words-only --unique || true)
 fi
+all_words=$(npx cspell --no-progress --no-summary --words-only --unique $(git ls-files | (grep -v '\.github/\.cspell/project-dictionary\.txt' || true)) || true)
+# TODO: handle SIGINT
+echo "${config_old}" >.cspell.json
 cat >.github/.cspell/rust-dependencies.txt <<EOF
 // This file is @generated by $(basename "$0").
 // It is not intended for manual editing.
 EOF
-if [[ -n "${words:-}" ]]; then
-    echo "${words}" >>.github/.cspell/rust-dependencies.txt
+if [[ -n "${dependencies_words:-}" ]]; then
+    echo $'\n'"${dependencies_words}" >>.github/.cspell/rust-dependencies.txt
 fi
+check_diff .github/.cspell/rust-dependencies.txt
 
-npx cspell --no-progress $(git ls-files)
+echo "+ npx cspell --no-progress --no-summary \$(git ls-files)"
+if ! npx cspell --no-progress --no-summary $(git ls-files); then
+    error "spellcheck failed: please fix uses of above words or add to .github/.cspell/project-dictionary.txt if correct"
+fi
 
-# Check duplications in dictionary
+# Make sure the project-specific dictionary does not contain duplicated words.
 for dictionary in .github/.cspell/*.txt; do
     if [[ "${dictionary}" == .github/.cspell/project-dictionary.txt ]]; then
         continue
     fi
     dup=$(sed '/^$/d' .github/.cspell/project-dictionary.txt "${dictionary}" | LC_ALL=C sort -f | uniq -d -i | (grep -v '//.*' || true))
     if [[ -n "${dup}" ]]; then
-        echo "error: duplicated words in dictionaries; please remove the following words from .github/.cspell/project-dictionary.txt"
+        error "duplicated words in dictionaries; please remove the following words from .github/.cspell/project-dictionary.txt"
         echo "======================================="
         echo "${dup}"
         echo "======================================="
     fi
 done
 
-# Fail CI if .github/.cspell/rust-dependencies.txt needs to be updated.
-if [[ -n "${CI:-}" ]]; then
-    if ! git --no-pager diff --exit-code .github/.cspell/rust-dependencies.txt; then
-        echo "error: please update .github/.cspell/rust-dependencies.txt"
+# Make sure the project-specific dictionary does not contain unused words.
+unused=''
+for word in $(grep -v '//.*' .github/.cspell/project-dictionary.txt || true); do
+    if ! grep <<<"${all_words}" -Eq -i "^${word}$"; then
+        unused+="${word}"$'\n'
     fi
+done
+if [[ -n "${unused}" ]]; then
+    error "unused words in dictionaries; please remove the following words from .github/.cspell/project-dictionary.txt"
```

Comment:
.github/.cspell/project-dictionary.txt　appears at least 6 times. I feel better if it is defined as a const variable. (If it's difficult, no problem!)